subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java

```diff
@@ -313,17 +294,50 @@ protected FileVisitResult doVisitFile(Path file, BasicFileAttributes attrs) {
             }
         }
 
+        @Nullable
+        private DirectorySnapshot followSymlink(Path file, String internedFileName) {
+            try {
+                Path targetDir = file.toRealPath();
+                String targetDirString = targetDir.toString();
+                if (!introducesCycle(targetDirString) && shouldVisitDirectory(targetDir, internedFileName)) {
+                    PathVisitor subtreeVisitor = new PathVisitor(
+                        predicate,
+                        hasBeenFiltered,
+                        hasher,
+                        stringInterner,
+                        defaultExcludes,
+                        collector,
+                        symbolicLinkMapping.withNewMapping(file.toString(), targetDirString, pathTracker)
+                    );
+                    Files.walkFileTree(targetDir, EnumSet.noneOf(FileVisitOption.class), Integer.MAX_VALUE, subtreeVisitor);
+                    return (DirectorySnapshot) subtreeVisitor.getResult();
+                } else {
+                    return null;
+                }
+            } catch (IOException e) {
+                throw new UncheckedIOException(String.format("Could not list contents of directory '%s'.", file), e);
+            }
+        }
+
         private boolean introducesCycle(String targetDirString) {
             return parentDirectories.contains(targetDirString);
         }
 
         private void visitResolvedFile(Path file, BasicFileAttributes targetAttributes, AccessType accessType) {
             String internedName = intern(file.getFileName().toString());
-            if (shouldVisit(file, internedName, false, pathTracker.getSegments())) {
+            if (shouldVisitFile(file, internedName)) {
                 builder.visitLeafElement(snapshotFile(file, internedName, targetAttributes, accessType));
             }
         }
 
+        private boolean shouldVisitDirectory(Path dir, String internedName) {
+            return pathTracker.isRoot() || shouldVisit(dir, internedName, true, pathTracker.getSegments());
+        }
+
+        private boolean shouldVisitFile(Path file, String internedName) {
+            return shouldVisit(file, internedName, false, pathTracker.getSegments());
```

Comment:
Nit: `shouldVisit()` always gets `pathTracker.getSegments()` passed in. It can look it up itself instead.

Reply:
Good point, done.